File: cracking_the_coding_interview/2_1_remove_dups/remove_dups.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <tuple>
#include <list>
#include <unordered_set>

// Remove all duplicates from given unsorted list.
//
// time complexity: O(N)
// space complexity: O(1)
void delete_dups(std::list<int>& list) {
    std::unordered_set<int> seen;

    auto it = list.begin();
    while(it != list.end()) {
        if(seen.count(*it) > 0) {
            it = list.erase(it);
        } else {
            seen.insert(*it);
            ++it;
        }
    }
}

// test implementation

std::ostream& operator<<(std::ostream& os, const std::list<int>& list) {
    os << "[";
    for (auto e: list) {
        os << e << ",";
    }
    os << "]";

    return os;
}


#define ASSERT(expression)                               \
    if(!(expression))                                    \
    {                                                    \
        std::cerr << "Assertion failed: " << #expression \
                  << "\nFile: " << __FILE__ << "\n"      \
                  << "\nLine: " << __LINE__ << "\n";     \
        return false;                                    \
    }

#define ASSERT_EQUAL(a, b)                                \
    if((a) != (b))                                        \
    {                                                     \
        std::cerr << "Assertion failed: " << #a << " == " \
                  << #b << "\n" << #a << " = " << (a)     \
                  << ", " << #b << " = " << (b)           \
                  << "\nFile: " << __FILE__ << "\n"       \
                  << "\nLine: " << __LINE__ << "\n";      \
        return false;                                     \
    }


bool run_tests() {
    using vec = std::vector<int>;

    std::vector<std::tuple<vec, vec>> test_cases;
    test_cases.emplace_back(vec{}, vec{});
    test_cases.emplace_back(vec{1}, vec{1});
    test_cases.emplace_back(vec{1, 2, 3}, vec{1, 2, 3});
    test_cases.emplace_back(vec{1, 1, 1}, vec{1});
    test_cases.emplace_back(vec{1, 1, 2, 2, 3, 3}, vec{1, 2, 3});
    test_cases.emplace_back(vec{1, 2, 3, 1, 2, 3}, vec{1, 2, 3});

    for (auto& [input, expected]: test_cases) {
        std::list<int> l1(input.begin(), input.end());
        std::list<int> l2(expected.begin(), expected.end());
        delete_dups(l1);
        ASSERT_EQUAL(l1, l2);
    }

    return true;
}


int main() {
    if(run_tests()) {
        std::cout << "C++ test succeeded\n";
        return 0;
    } else {
        std::cout << "Failure in C++ tests\n";
        return 1;
    }
}

File: cracking_the_coding_interview/2_1_remove_dups/remove_dups.rs
use std::collections::{HashSet, LinkedList};
use std::iter::FromIterator;

// Remove all duplicates from given unsorted list.
//
// time complexity: O(N)
// space complexity: O(1)
fn delete_dups(list: &mut LinkedList<i32>) {
    let mut seen = HashSet::new();
    let mut current = list.pop_front();

    while let Some(val) = current {
        if seen.insert(val) {
            // If the value is inserted, it is not a duplicate.
            list.push_back(val);
        }
        current = list.pop_front();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_delete_dups() {
        let test_cases = vec![
            (vec![], vec![]),
            (vec![1], vec![1]),
            (vec![1, 2, 3], vec![1, 2, 3]),
            (vec![1, 1, 1], vec![1]),
            (vec![1, 1, 2, 2, 3, 3], vec![1, 2, 3]),
            (vec![1, 2, 3, 1, 2, 3], vec![1, 2, 3]),
        ];

        for (input, expected) in test_cases {
            let mut list = LinkedList::from_iter(input);
            let expected_list = LinkedList::from_iter(expected);
            delete_dups(&mut list);
            assert_eq!(list, expected_list);
        }
    }
}
